### ISSUE TITLE
Unable to cast object of type 'Microsoft.CodeAnalysis.CSharp.Symbols.SubstitutedTypeParameterSymbol' to type 'Microsoft.CodeAnalysis.CSharp.Symbols.NamedTypeSymbol'

### ISSUE THREAD

**Issue by jamesfoster** [dotnet/roslyn#37269]
**Version Used**: Microsoft (R) Visual C# Compiler version 3.100.19.22309 (ce8ac3f7)

Visual Studio crashes when trying to type out the following. While there are compile errors (you haven't finished typing) VS is very sluggish and then crashes.

```
    public static class ResultExtensions
    {
        public static async Task<Result<TResult, B>> MapAsync<A, B, TResult>(
            this Result<A, B> result,
            Func<A, Task<TResult>> map
        )
        {
            return await result
               .Match(
                    success: async task => Result.Success<TResult, B>(await map(task)),
                    failure: value => Task.FromResult(Result.Failure<TResult, B>(value))
                );
        }
    }
```

I can reliably cause the crash by, for example, replacing `Result.Success<TResult, B>` (line 10) with `Result.Success<TResult>`.

Below are the signatures of the supporting code:
```
  public class Result<A, B> {
    public TResult Match<TResult>(Func<A, TResult> success, Func<B, TResult> failure);
  }
  public static class Result {
    public Result<A, B> Success<A, B>(A value) => ...
    public Result<A, B> Failure<A, B>(B value) => ...
  }
```

Stack trace from Event Viewer:
```
Application: devenv.exe
Framework Version: v4.0.30319
Description: The application requested process termination through System.Environment.FailFast(string message).
Message: System.InvalidCastException: Unable to cast object of type 'Microsoft.CodeAnalysis.CSharp.Symbols.SubstitutedTypeParameterSymbol' to type 'Microsoft.CodeAnalysis.CSharp.Symbols.NamedTypeSymbol'.
   at Microsoft.CodeAnalysis.CSharp.NullableWalker.TryGetReturnType(TypeWithAnnotations& type)
   at Microsoft.CodeAnalysis.CSharp.NullableWalker.VisitReturnStatementNoAdjust(BoundReturnStatement node)
   at Microsoft.CodeAnalysis.CSharp.AbstractFlowPass`1.VisitReturnStatement(BoundReturnStatement node)
   at Microsoft.CodeAnalysis.CSharp.BoundReturnStatement.Accept(BoundTreeVisitor visitor)
   at Microsoft.CodeAnalysis.CSharp.AbstractFlowPass`1.VisitWithStackGuard(BoundNode node)
   at Microsoft.CodeAnalysis.CSharp.AbstractFlowPass`1.VisitAlways(BoundNode node)
   at Microsoft.CodeAnalysis.CSharp.AbstractFlowPass`1.Visit(BoundNode node)
   at Microsoft.CodeAnalysis.CSharp.NullableWalker.VisitStatement(BoundStatement statement)
   at Microsoft.CodeAnalysis.CSharp.AbstractFlowPass`1.VisitStatements(ImmutableArray`1 statements)
...
```

Full stack trace: https://pastebin.com/raw/1QY6hnEx

**Comment by cston** [comment]
@jamesfoster, thanks for reporting this issue and for providing the repro!

**Comment by jamesfoster** [comment]
You're welcome, thanks for fixing it so quickly :D I look forward to the next update.

**Comment by jamesfoster** [comment]
@cston Before I raise a new issue. Do you think that this will have been fixed by the same change? It's a different exception but it's related to the same "Type Gymnastics"

I have 2 methods:

```csharp
public static Result<TResult, B> FlatMap<A, B, TResult>(
   this Result<A, B> result,
   Func<A, Result<TResult, B>> map
)

// and

public static Task<Result<TResult, B>> FlatMapAsync<A, B, TResult>(
   this Result<A, B> result,
   Func<A, Task<Result<TResult, B>>> map
)
```

When I type the following I get the below crash.

```csharp
Func<int, Task<long>> f;
Result<int, string> result;
Result<long, string> next = await result.FlatMap(async value => Result.Success(await f(value)));
```

In the above I'm calling the wrong method. It should be (of course) `FlatMapAsync`

```
Message: System.InvalidOperationException: Sequence contains more than one element
   at System.Linq.Enumerable.Single[TSource](IEnumerable`1 source)
   at Microsoft.CodeAnalysis.CSharp.NullableWalker.TryGetReturnType(TypeWithAnnotations& type)
   at Microsoft.CodeAnalysis.CSharp.NullableWalker.VisitReturnStatementNoAdjust(BoundReturnStatement node)
```

Full stacktrace: https://pastebin.com/raw/b6P9R0mQ

**Comment by jamesfoster** [comment]
I'm resorting to writing my code in comments then uncommenting it. It's very painful to work like this.

**Comment by cston** [comment]
@jamesfoster, the exception is distinct but the cause is the same. I have verified this case is fixed with #37488. Thanks for following up.
